index.bs

```diff
@@ -9,8 +9,184 @@ Editor: Kent Tamura, Google https://www.google.com/, tkent@google.com
+over time depending on user interaction and other factors, and are exposed to
+web authors through pseudo classes. For example, some form controls have the
+“invalid” state, which is exposed through the '':invalid'' [=pseudo-class=].
+
+Like built-in elements, [=custom elements=] can have various states to be in too,
+and [=custom element=] authors want to expose these states in a similar fashion
+as the built-in elements.
+
+## Solution ## {#solution}
+
+This specification defines an API to inform [=custom element=]'s states to the
+user agent, and a [=pseudo-class=] to select elements with specific states.
+The former is the {{ElementInternals/states}} IDL attribute of
+{{ElementInternals}}, and the latter is the '':state()'' [=pseudo-class=].
+
+
+<div class="example" id="ex-intro">
+
+Assume that <code>LabeledCheckbox</code> doesn't expose its "checked" state
+via a content attribute.
+
+<!-- <xmp> doesn't work well with backqoutes.-->
+<pre class="lang-html">
+&lt;!DOCTYPE html>
+&lt;body>
+&lt;!-- Basic usage: -->
+&lt;script>
+class LabeledCheckbox extends  HTMLElement {
+  constructor() {
+    super();
+    this._internals = this.attachInternals();
+    this.addEventListener('click', this._onClick.bind(this));
+
+    const shadowRoot = this.attachShadow({mode: 'closed'});
+    shadowRoot.innerHTML =
+      &#96;&lt;style>
+       :host::before {
+         content: '[ ]';
+         white-space: pre;
+         font-family: monospace;
+       }
+       :host(:state(checked))::before { content: '[x]' }
+       &lt;/style>
+       &lt;slot>Label&lt;/slot>&#96;;
+  }
+
+  get checked() { return this._internals.states.contains('checked'); }
+
+  set checked(flag) {
+    this._internals.states.toggle('checked', !!flag);
+  }
+
+  _onClick(event) {
+    this.checked = !this.checked;
+  }
+}
+
+customElements.define('labeled-checkbox', LabeledCheckbox);
+&lt;/script>
+
+&lt;style>
+labeled-checkbox { border: dashed red; }
+labeled-checkbox:state(checked) { border: solid; }
+&lt;/style>
+
+&lt;labeled-checkbox>You need to check this&lt;/labeled-checkbox>
+
+<!-- Works even on ::part()s -->
+&lt;script>
+class QuestionBox extends HTMLElement {
+  constructor() {
+    super();
+    const shadowRoot = this.attachShadow({mode: 'closed'});
+    shadowRoot.innerHTML =
+      &#96;&lt;div>&lt;slot>Question&lt;/slot>&lt;/div>
+       &lt;labeled-checkbox part='checkbox'>Yes&lt;/labeled-checkbox>&#96;;
+  }
+}
+customElements.define('question-box', QuestionBox);
+&lt;/script>
+
+&lt;style>
+question-box::part(checkbox) { color: red; }
+question-box::part(checkbox):state(checked) { color: green; }
+&lt;/style>
+
+&lt;question-box>Continue?&lt;/question-box>
+&lt;/body>
+</pre>
+</div>
+
+
+Exposing custom element states {#exposing}
+============================
+
+Each <a>autonomous custom element</a> has <dfn>custom states</dfn>, a [=/set=]
+of strings. It's an empty set initially.
```

Comment:
I think we do not want to have a "custom states" internal slot owned by the element. Instead we want to have a DOMTokenList internal slot (e.g. "states token list") owned by the element, and use that DOMTokenList's token set.

Reply:
Done.